File: python/backtrack.py
def backtrack(path, used, N):
    # 종료 조건: path 길이가 N이 되면 하나의 순열이 완성된 것이므로 출력
    if len(path) == N:
        print("결과:", path)
        return

    # 1부터 N까지의 숫자 중에서 하나씩 선택
    for i in range(1, N+1):
        if not used[i]:
            print(f"숫자 {i} 선택 전: {path}")  # 선택 전 상태 출력

            used[i] = True        # 숫자 i를 사용했다고 표시
            print("(1) --->", used)
            path.append(i)        # 현재 경로(path)에 i 추가
            print(f"숫자 {i} 선택 후: {path}")  # 선택 후 상태 출력

            print(f"숫자 {i} 백트래킹 전: {path}")  # 백트래킹 전 상태 출력
            backtrack(path, used, N)  # 재귀 호출로 다음 숫자 선택
            path.pop()            # 마지막에 추가한 숫자 제거 (백트래킹)
            used[i] = False       # 숫자 i를 다시 사용 가능하도록 되돌림
            print("(2) --->", used)
            print(f"숫자 {i} 백트래킹 후: {path}")  # 백트래킹 후 상태 출력

# N=3이면 1부터 3까지의 숫자로 만들 수 있는 순열 출력
N = 3
backtrack([], [False] * (N+1), N)

File: java/Backtrack.java
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

public class Backtrack {
    public static void main(String[] args) {
        int N = 3;
        boolean[] used = new boolean[N + 1]; // 인덱스 1~N 사용
        List<Integer> path = new ArrayList<>();
        backtrack(path, used, N);
    }

    public static void backtrack(List<Integer> path, boolean[] used, int N) {
        if (path.size() == N) {
            System.out.println("결과: " + path);
            return;
        }


        for (int i = 1; i <= N; i++) {
            if (!used[i]) {
                System.out.println("숫자 " + i + " 선택 전: " + path);

                used[i] = true; // 숫자 i 사용 표시
                System.out.println("(1) ---> " + Arrays.toString(used));
                path.add(i); // path에 추가
                System.out.println("숫자 " + i + " 선택 후: " + path);

                System.out.println("숫자 " + i + " 백트래킹 전: " + path);
                backtrack(path, used, N); // 재귀호출
                path.remove(path.size() - 1); // 백트래킹
                used[i] = false; // 백트래킹: 사용 취소
                System.out.println("(2) ---> " + Arrays.toString(used));
                System.out.println("숫자 " + i + " 백트래킹 후: " + path);
            }
        }
    }
}
